modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java

```diff
@@ -0,0 +1,872 @@
+        //      1) column from sub-query table
+        //      2) can be accessible by any AND branch of the WHERE tree:
+        //                  WHERE
+        //                    |
+        //                   AND
+        //                  /    \
+        //                AND    OR
+        //               /   \   / \
+        //              c1   c2 *   * // * - can't be reached by AND branch
+        //
+        //      2) is strict predicate to sub-query table (only equality supported)
+        //      3) this set is superset for any of unique index, e.g. for query like
+        //                 WHERE c1 = v1
+        //                   AND c2 = v2
+        //                         ...
+        //                   AND ci = vi
+        //                   AND ci+1 = vi+1
+        //                   AND ci+2 = vi+2,
+        //         there is unique index unique_idx(c1, c2, ..., ci)
+        //
+        // Nota bene: equality should be only between columns of different tables
+        // (or between column and constant\param), because condition like 'PK = PK'
+        // filter out nothing.
+
+        if (subS.where() == null)
+            return false;
+
+        ASTNodeFinder opEqFinder = new ASTNodeFinder(
+            subS.where(),
+            (parent, child) -> ELEMENT_IS_EQ.test(parent),
+            ELEMENT_IS_AND_OPERATION.or(ELEMENT_IS_EQ)
+        );
+
+        List<GridSqlColumn> sqlCols = new ArrayList<>();
+        ASTNodeFinder.Result res;
+        while ((res = opEqFinder.findNext()) != null) {
+            GridSqlOperation op = (GridSqlOperation)res.getEl();
+
+            assert op.size() == 2;
+
+            GridSqlColumn[] colArr = new GridSqlColumn[2];
+
+            for (int i = 0; i < 2; i++) {
+                if (op.child(i) instanceof GridSqlColumn)
+                    colArr[i] = op.child(i);
+                else if (op.child(i) instanceof GridSqlOperation
+                    && ((GridSqlOperation)op.child(i)).operationType() == NEGATE
+                    && op.child() instanceof GridSqlColumn)
+                    colArr[i] = op.child(i).child();
+            }
+
+            if (colArr[0] == null || colArr[1] == null
+                // reference equality used on purpose here, because we should to get known
+                // whether the column refers to exactly the same table
+                || colArr[0].expressionInFrom() != colArr[1].expressionInFrom()
+            ) {
+                if (colArr[0] != null && colArr[0].expressionInFrom() == subS.from())
+                    sqlCols.add(colArr[0]);
+
+                if (colArr[1] != null && colArr[1].expressionInFrom() == subS.from())
+                    sqlCols.add(colArr[1]);
+            }
+        }
+
+        GridSqlAst subTbl = GridSqlAlias.unwrap(subS.from());
+
+        return subTbl instanceof GridSqlTable && isUniqueIndexExists(sqlCols, (GridSqlTable)subTbl);
```

Comment:
@Test
    public void test223() {
        String outerSqlTemplate = "select e.id, e.name from emp e where exists (%s) and e.name = ''";
        String subSql = "select 1 from dep d, dep2 d2 where d.id = d2.id and d.id = d2.id2 and d.id = e.id";
        subSql = "select 1 from dep2 d2 where (d2.name = '' and d2.id = e.id) or (d2.name = '' and d2.id2 = e.id)";
        String resSql = String.format(outerSqlTemplate, subSql);
        check(resSql, 2);
    }

Reply:
testOptimizationShouldNotBeApplied2